### PR TITLE
assistant tests: Testing hot-reload

### PR DESCRIPTION
**Pull request by robin-da**
This adds a test to the assistant integration tests for testing the
hot-reload feature.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTests.hs [hunk 1]
```diff
@@ -264,6 +265,111 @@ packagingTests = testGroup "packaging"
...
+              , "--json-api-port"
+              , show jsonApiPort
+              ]) {std_in = CreatePipe, cwd = Just projDir}
+        withCurrentDirectory projDir $
+          withCreateProcess startProc $ \startStdIn _ _ startPh ->
+            race_ (waitForProcess' startProc startPh) $
+             do
+              -- The hard-coded secret for testing is "secret".
+              let token =
+                    JWT.encodeSigned
```
Comment (hurryabit) [inline daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTests.hs:314]: Don't we have a helper function for creating such tokens in one of the tests you wrote recently?

File: daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTests.hs [hunk 2]
```diff
@@ -264,6 +265,111 @@ packagingTests = testGroup "packaging"
...
+              statusCode (responseStatus queryResponse) @?= 200
+              case Aeson.decode (responseBody queryResponse) of
+                Just (Aeson.Object body)
+                  | Just (Aeson.Array result) <- HashMap.lookup "result" body ->
+                    case Vector.toList result of
+                      [Aeson.Object r]
+                        | Just (Aeson.Object payload) <- HashMap.lookup "payload" r ->
+                          HashMap.lookup "newFieldName" payload @?= Just "Alice"
+                      other -> assertFailure $ "Bad result in response: " <> show other
+                other -> assertFailure $ "Expected JSON object in response body: " <> show other
```
Comment (hurryabit) [inline daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTests.hs:369]: Would it simplify things if built an aeson parser here? If so, let's do that in a separate PR.

File: daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTests.hs [hunk 3]
```diff
@@ -264,6 +265,111 @@ packagingTests = testGroup "packaging"
...
+                  , "import Daml.Script"
+                  , "template S with newFieldName : Party where signatory newFieldName"
+                  , "init : Script ()"
+                  , "init = do"
+                  , "  [aliceDetails] <- listKnownParties"
+                  , "  let alice = party aliceDetails"
+                  , "  alice `submit` createCmd (S alice)"
+                  , "  pure ()"
+                  ]
+              maybe (fail "No start process stdin handle") (\h -> hPutChar h 'r' >> hFlush h) startStdIn
```
Comment (hurryabit) [inline daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTests.hs:347]: Will this work on Windows? Are do we need to send a separate ENTER press?

**Review by hurryabit — APPROVED** [review]
Very nice! Thank you.

Would it be possible to test that the codegen has run as well? If yes, let's do it in a followup PR.

File: daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTests.hs (hunk 3 above)
Comment (robin-da) [reply daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTests.hs:347]: It should work because of the extra `flush`. The extra `ENTER` just forces the flush on Windows, but we're not checking for an extra `ENTER` to start the rebuild on Windows.

File: daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTests.hs (hunk 1 above)
Comment (robin-da) [reply daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTests.hs:314]: The new http json api tests create the sandbox and http json api with a configurable secret. But there is no function exported for creating a token. I think the Assistant/IntegrationTests.hs are quite messy right now and would benefit from factoring out a few commonly used functions such as the token creation.

File: daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTests.hs (hunk 2 above)
Comment (robin-da) [reply daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTests.hs:369]: yes, definitely clearer. Let's do a cleanup PR that adds some clarity to the assistant integration tests.

File: daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTests.hs (hunk 1 above)
Comment (hurryabit) [reply daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTests.hs:314]: Ok, thanks for the clarification. Do you have some concrete ideas for such refactorings? If so, please go ahead. If not, let's chat about it.

File: daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTests.hs (hunk 3 above)
Comment (hurryabit) [reply daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTests.hs:347]: Ah, that makes sense. Thanks.
